lib/firmata.js

```diff
@@ -1411,6 +1459,178 @@ Board.prototype.stepperStep = function(deviceNum, direction, steps, speed, accel
+ */
+
+Board.prototype.serialConfig = function(options) {
+
+  var portId;
+  var baud;
+  var bytesToRead;
+  var rxPin;
+  var txPin;
+
+  if (typeof options === "object" && options !== null) {
+    portId = options.portId;
+    baud = options.baud;
+    bytesToRead = options.bytesToRead;
+    rxPin = options.rxPin;
+    txPin = options.txPin;
+  }
+
+  // TODO - if info about the Serial capabilities can be added to the capability query, a
+  // reasonable default value can be added and this error can be removed. Until then it is
+  // not possible to know if a board supports only HW serial or only SW serial or both so
+  // there is really no way to add a default portId value.
+  if (typeof portId === "undefined") {
+    throw new Error("portId must be specified, see SERIAL_PORT_IDs for options.");
+  }
+
+  baud = baud || 57600;
+  bytesToRead = bytesToRead || 0;
+
+  var data = [
+    START_SYSEX,
+    SERIAL_MESSAGE,
+    SERIAL_CONFIG | portId,
+    baud & 0x007F,
+    (baud >> 7) & 0x007F,
+    (baud >> 14) & 0x007F,
+    bytesToRead & 0x007F,
+    (bytesToRead >> 7) & 0x007F
+  ];
+  if (rxPin) {
+    data.push(rxPin);
+  }
+  if (txPin) {
+    data.push(txPin);
+  }
+  data.push(END_SYSEX);
+  this.transport.write(new Buffer(data));
+};
+
+/**
+ * Write an array of bytes to the specified serial port.
+ * @param {number} portId The serial port to write to.
+ * @param {array} inBytes An array of bytes to write to the serial port.
+ */
+
+Board.prototype.serialWrite = function(portId, inBytes) {
+  var data = [
+    START_SYSEX,
+    SERIAL_MESSAGE,
+    SERIAL_WRITE | portId
+  ];
+  for (var i = 0, len = inBytes.length; i < len; i++) {
+    data.push(inBytes[i] & 0x007F);
+    data.push((inBytes[i] >> 7) & 0x007F);
+  }
+  data.push(END_SYSEX);
+  if (data.length > 0) {
+    this.transport.write(new Buffer(data));
+  }
+};
+
+/**
+ * Start continuous reading of the specified serial port. The port is checked for data each
+ * iteration of the main Arduino loop.
+ * @param {number} portId The serial port to start reading continuously.
+ * @param {function} callback A function to call when we have received the bytes.
+ */
+
+Board.prototype.serialRead = function(portId, callback) {
+  var data = [
+    START_SYSEX,
+    SERIAL_MESSAGE,
+    SERIAL_READ | portId,
+    this.SERIAL_MODES.READ_CONTINUOUS,
+    END_SYSEX
+  ];
+  this.transport.write(new Buffer(data));
+
+  this.on("serial-data-" + portId, callback);
+};
+
+/**
+ * Stop continuous reading of the specified serial port. This does not close the port, it stops
+ * reading it but keeps the port open.
+ * @param {number} portId The serial port to stop reading.
+ */
+
+Board.prototype.serialStop = function(portId) {
+  var data = [
+    START_SYSEX,
+    SERIAL_MESSAGE,
+    SERIAL_READ | portId,
+    this.SERIAL_MODES.STOP_READING,
+    END_SYSEX
+  ];
+  this.transport.write(new Buffer(data));
+
+  this.removeAllListeners(["serial-data-" + portId]);
```

Comment:
@rwaldron is this the right approach here? I could alternatively accept an optional callback param on line 1567 and if passed call `this.removeListener("serial-data-" + portId, callback);`